### DISCUSSION TITLE
Could not build with CUDA_ARCH_LIST=Common

### DISCUSSION THREAD

**Discussion by hubutui** [apache/incubator-mxnet#16852] [Q&A] [answered]
## Description
I could not build with `CUDA_ARCH_LIST=Common` on ArchLinux.
### Error Message
[mxnet-git.log](https://github.com/apache/incubator-mxnet/files/3863306/mxnet-git.log)

## To Reproduce
```shell
  cmake_opts=(
    -DBUILD_CPP_EXAMPLES:BOOL=OFF
    -DBUILD_TESTING:BOOL=OFF
    -DCMAKE_BUILD_TYPE:String=Release
    -DCMAKE_EXE_LINKER_FLAGS="$(pkg-config --libs cblas)"
    -DCMAKE_INSTALL_LIBDIR:PATH=lib
    -DCMAKE_INSTALL_PREFIX:PATH=/usr
    -DCMAKE_SHARED_LINKER_FLAGS="$(pkg-config --libs cblas)"
    -DCMAKE_SKIP_INSTALL_RPATH:BOOL=ON
    -DINSTALL_DOCUMENTATION:BOOL=OFF
    -DUSE_BLAS=open
    -DUSE_CPP_PACKAGE:BOOL=ON
    -DUSE_DIST_KVSTORE:BOOL=ON
    -DUSE_GPERFTOOLS:BOOL=OFF
    -DUSE_JEMALLOC:BOOL=OFF
    -DUSE_LAPACK:BOOL=ON
    -DUSE_MKL_IF_AVAILABLE:BOOL=OFF
    -DUSE_OPENMP:BOOL=ON
    -DUSE_S3:BOOL=ON
)
  cmake \
    ${cmake_opts[@]} \
    -DCUDA_ARCH_LIST=${cudaarch} \
    -DUSE_CUDA:BOOL=ON \
    -DUSE_CUDNN:BOOL=ON \
    -DUSE_MKLDNN:BOOL=ON \
    -DUSE_NCCL:BOOL=ON \
    -DUSE_OPENCV:BOOL=OFF \
    ..
  make
  cd ../python
  python setup.py build --with-cython
```

## Environment info
Same as https://github.com/apache/incubator-mxnet/issues/15263#issue-457222298

## What have you tried to solve it?

1. It's OK to build with `CUDA_ARCH_LIST=Pascal`
2. I try https://github.com/apache/incubator-mxnet/issues/15971#issuecomment-526765550, and add `-mcmodel=large` to both `CMAKE_C_FLAGS` and `CMAKE_CXX_FLAGS`, but with no luck.
3. https://github.com/apache/incubator-mxnet/issues/14263#issuecomment-541334739 mentioned a solution, but I'm not sure how to do with cmake building.

**Comment by samskalicky** [comment]
@mxnet-label-bot update [Question, Build]

**Comment by hubutui** [comment]
I also check https://github.com/apache/incubator-mxnet/pull/15981, but it seems it's not fixed yet. According to https://github.com/apache/incubator-mxnet/blob/master/CMakeLists.txt#L51, `ENABLE_TESTCOVERAGE=OFF` by default.

**Comment by leezu** [comment]
Try building `binutils` locally with `--enable-64-bit-archive`.
https://github.com/apache/incubator-mxnet/issues/14263#issuecomment-564859500

**Comment by hubutui** [comment]
@leezu Hey, I try rebuilt `binutils` locally with `--enable-64-bit-archive`, and then build mxnet. It didn't work. I could confirm that `CMAKE_AR=/usr/bin/ar`, and this is the newly build binutils.

**Comment by leezu** [comment]
@hubutui besides using the 64 bit ar with `CMAKE_AR` and `-mcmodel=medium`, you can also add  `-Wl,--no-relax`. In my case, the error message instructed to specify `-Wl,--no-relax`.

Also, for cmake, I find that using `cmake -GNinja` to select ninja build system makes the relocation truncation issue less likely

**Comment by hubutui** [comment]
You mean I could try with `-mcmodel=large -Wl --no-relax` or `-mcmodel=medium -Wl --no-relax` again?

**Comment by leezu** [comment, marked as answer]
```
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mcmodel=medium -Wl,--no-relax")
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mcmodel=medium -Wl,--no-relax")
```
the `,` is required

**Comment by hubutui** [comment]
@leezu You mean just append these two line to `CMakeLists.txt`? I have try it, but with no lock. I'm currently manual set `CUDA_COMMON_GPU_ARCHITECTURES` to compute capability >= 3.7.
